src/crimson/os/seastore/seastore_types.h

```diff
@@ -1006,33 +1007,64 @@ constexpr journal_seq_t JOURNAL_SEQ_MAX{
 // JOURNAL_SEQ_NULL == JOURNAL_SEQ_MAX == journal_seq_t{}
 constexpr journal_seq_t JOURNAL_SEQ_NULL = JOURNAL_SEQ_MAX;
 
+// logical offset, see LBAManager, TransactionManager
+using extent_len_t = uint32_t;
+constexpr extent_len_t EXTENT_LEN_MAX =
+  std::numeric_limits<extent_len_t>::max();
+
+using extent_len_le_t = ceph_le32;
+inline extent_len_le_t init_extent_len_le(extent_len_t len) {
+  return ceph_le32(len);
+}
+
+using shard_t = int8_t;
+using pool_t = int64_t;
+// Note: this is the reversed version of the object hash
+using crush_hash_t = uint32_t;
+using snap_t = uint64_t;
+using gen_t = uint64_t;
+static_assert(sizeof(shard_t) == sizeof(ghobject_t().shard_id.id));
+static_assert(sizeof(pool_t) == sizeof(ghobject_t().hobj.pool));
+static_assert(sizeof(crush_hash_t) == sizeof(ghobject_t().hobj.get_bitwise_key_u32()));
+static_assert(sizeof(snap_t) == sizeof(ghobject_t().hobj.snap.val));
+static_assert(sizeof(gen_t) == sizeof(ghobject_t().generation));
+
 // logical addr, see LBAManager, TransactionManager
 class laddr_t {
 public:
   // the type of underlying integer
   using Int = uint64_t;
 
+  // laddr_t is block aligned, one logical address represents one 4KiB block in disk
+  static constexpr unsigned UNIT_SHIFT = 12;
+  static constexpr unsigned UNIT_SIZE = 1 << UNIT_SHIFT; // 4096
+  static constexpr unsigned UNIT_MASK = UNIT_SIZE - 1;
+
   constexpr laddr_t() : laddr_t(0) {}
-  constexpr explicit laddr_t(Int value) : value(value) {}
 
-  bool is_aligned(Int alignment) const {
-    assert(alignment != 0);
-    assert((alignment & (alignment - 1)) == 0);
-    return value == p2align(value, alignment);
+  static constexpr laddr_t from_byte_offset(Int value) {
+    if (!std::is_constant_evaluated()) {
```

Comment:
It's just the C++20 version of `if consteval` which requires C++23. `assert` has side effects and can't be used in const expression context.

Reply:
Seems to me `std::is_constant_evaluated()` will cause the assert mostly be ignored. So, how about making this method `inline` instead of more strict `constexpr` ?

Reply:
> Seems to me `std::is_constant_evaluated()` will cause the assert mostly be ignored.

It will be executed at runtime, like this:
```cpp
#include <type_traits>

constexpr int foo(int i) {
  if (!std::is_constant_evaluated()) {
    assert(i != 0);
  }
  return i;
}

int main() {
  constexpr int i = foo(0); // pass at compile time
  int j = foo(0); // fail at run time
}
```

> So, how about making this method `inline` instead of more strict `constexpr` ?

The constexpr laddr_t in test_transaction_manager.cc will also be removed.

Reply:
> Seems the parameter of constexpr function is not the const expression, I didn't found other ways to validate it.

Do you mean it is mostly not a constexpr expression at runtime? If true, let's simply convert it to a normal method without `constexpr`, what do you think?

Reply:
Here `const expression` should be `constant expression`, I mean we can't write code like `static_assert(offset & UNIT_MASK != 0)` here, thus `if (!std::is_constant_evaluated())` is just to make the `assert` is not evaluated at compile time.

Reply:
That's right. I meant it's fine to drop `constexpr` from `from_byte_offset()` to enforce necessary asserts in the debug build.